recap/readers/hive/metastore.py

```diff
@@ -0,0 +1,205 @@
+from thrift_gen import ThriftHiveMetastore as hms
+from thrift_gen.ttypes import *
+from thrift.transport import TTransport, TSocket
+from thrift.protocol.TBinaryProtocol import TBinaryProtocol
+from enum import Enum
+from typing import Dict, List
+
+class HMS:
+    def __init__(self, host='localhost', port=9090):
+        self.host = host
+        self.port = port
+        self.socket = TSocket.TSocket(self.host, self.port)
+        self.transport = TTransport.TBufferedTransport(self.socket)
+        self.protocol = TBinaryProtocol(self.transport)
+        self.client = hms.Client(self.protocol)
+
+    def connect(self):
+        self.transport.open()
+
+    def disconnect(self):
+        self.transport.close()
+
+    def get_databases(self):
+        return self.client.get_all_databases()
+        
+    def get_tables(self, database):
+        return self.client.get_all_tables(database)
+        
+    def get_table(self, database, table):
+        return self.client.get_table(database, table)
+        
+    def get_catalogs(self):
+        return self.client.get_catalogs()
+        
+    def get_schema(self, database, table):
+        return self.client.get_schema(database, table)
+        
+    def get_fields(self, database, table):
+        return self.client.get_fields(database, table)
+    
+class PrincipalType(Enum):
+    ROLE = "ROLE"
+    USER = "USER"
+
+class PrimitiveCategory(Enum):
+    VOID = "VOID"
+    BOOLEAN = "BOOLEAN"
+    BYTE = "BYTE"
+    SHORT = "SHORT"
+    INT = "INT"
+    LONG = "LONG"
+    FLOAT = "FLOAT"
+    DOUBLE = "DOUBLE"
+    STRING = "STRING"
+    DATE = "DATE"
+    TIMESTAMP = "TIMESTAMP"
+    TIMESTAMPLOCALTZ = "TIMESTAMPLOCALTZ"
+    BINARY = "BINARY"
+    DECIMAL = "DECIMAL"
+    VARCHAR = "VARCHAR"
+    CHAR = "CHAR"
+    INTERVAL_YEAR_MONTH = "INTERVAL_YEAR_MONTH"
+    INTERVAL_DAY_TIME = "INTERVAL_DAY_TIME"
+    UNKNOWN = "UNKNOWN"
+
+class HTypeCategory(Enum):
+    PRIMITIVE = PrimitiveCategory
+    STRUCT = "STRUCT"
+    MAP = "MAP"
+    LIST = "LIST"
+    UNION = "UNION"
+
+class HType:
```

Comment:
Curious why you're creating a new HiveType vs. using `recap.types` directly?

Reply:
That's a great question and something that was bothering and still bothers me. The Thrift types are too low level in many cases, e.g. the type you get back for the PrincipalType is an int that then you use to instantiate the actual type. 

It feels to me that there's a lot of low level complexity there that could / should be hidden from anyone who's building on top of the Hive Metastore. That's why I added this modeling layer for the types. This is also inspired by how Trino is doing it btw in their implementation, I trust their implementation because as a federated query engine they always needed to figure out the right abstractions.

Also, I might be misinterpreting something here in terms of the Recap design. We have Readers and converters. I operated with the assumption that a reader provides access to a "domain" and then there's a converter that turns that into the Recap type system. Is this a wrong assumption?

Reply:
> Also, I might be misinterpreting something here in terms of the Recap design. We have Readers and converters. I operated with the assumption that a reader provides access to a "domain" and then there's a converter that turns that into the Recap type system. Is this a wrong assumption?

More or less.

* Converter: given a schema, returns Recap types. The formats are usually text-based (proto, avro, json schema, etc).
* Reader: Reads a schema from a registry/catalog and returns a StructType.

Readers usually use converters to take the schema that it reads and convert it to a Recap StructType before returning it.

The reason for the distinction is that many different registries/catalogs all have the same schema formats (proto, avro, json schema, thrift). This way the DatahubReader, ConfluentRegistryReader, etc can all use the same converters.

Reply:
TBH, I think your HType approach is fine. I've been noodling about breaking the ThriftHiveMetastore stuff you're doing out into a separate pypi package, repo, and project anyway. It seems useful beyond Recap to have a Python-based Hive Metastore API that uses the Thrift interface.

Basically, this library:

https://github.com/quintoandar/hive-metastore-client

But with all the Thrift APIs supported (what you're doing).

Reply:
Yep, I agree that there's value in breaking this into a separate package. Let's chat about how it should be done and I'm happy to take that direction.

Reply:
Want me to set up a stub project in recap-build/pyhive-metastore ?

Reply:
That would be awesome! set up the stub and I'll work on that.

Reply:
I've opened:

https://github.com/recap-build/pymetastore

You're set up as a maintainer. :)